### PR TITLE
✨ Spin flavor clusters as tilt resources

### LINKED ISSUE #542: Make each of the flavors in available to deploy in Tilt
/kind feature

**Describe the solution you'd like**
[A clear and concise description of what you want to happen.]

When one runs `make tilt-up`, you are greeted with a Tilt console and browser window running a kind CAPZ management cluster, but there is no workload cluster running. It would be cool to be able to deploy any of the flavors under `./templates/cluster-template*.yaml` as a [local_resource](https://docs.tilt.dev/local_resource.html) via Tilt. It would give a nice visual experience for developers hacking on CAPZ.

As part of deploying one of the flavors, the tilt-settings file should probably offer overrides for environment variables which are included in the flavors, and provide sane default for the vars that are currently required (AZURE_LOCATION, CLUSTER_NAME, etc).

### PR DESCRIPTION
**Pull request by jroden**
parent 5a7bedf05cb574a7825608561192d35b535abb9c
author James Oden <joden1995@gmail.com> 1591073777 -0400
committer James Oden <joden1995@gmail.com> 1591153490 -0400

✨ Spin flavor clusters as tilt resources

**What this PR does / why we need it**:
This is for feature #542.  This allows users to spin up worker clusters of different flavors that are represented by tilt as local resources.

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #542

**Special notes for your reviewer**:
This is my first PR, thanks for the help :) 

_Please confirm that if this PR changes any image versions, then that's the sole change this PR makes._

**Release note**:

```release-note
Users can now spin up worker clusters of a particular flavor using Tilt, with the clusters being represented as a local resource.  Users can also specify custom variable overrides for the clusters in tilt-settings.json
```

### CODE REVIEW

File: templates/flavors/README.md [hunk 1]
```diff
@@ -14,3 +14,77 @@ appended to the end of the cluster-template.yaml, e.g cluster-template-{director
 flavor can be used by specifying `--flavor {directory-name}`.
 
 To generate all CAPZ flavors, run `make generate-flavors`.
+
+
+## Running flavor clusters as a tilt resource
```
Comment (CecileRobertMichon) [inline templates/flavors/README.md:19]: let's put a link to these instructions in https://github.com/kubernetes-sigs/cluster-api-provider-azure/blob/master/docs/development.md#using-tilt

Comment (jroden) [reply templates/flavors/README.md:19]: Thanks for the help!  I just added new commit for this one.  Let me know your thoughts!

File: templates/flavors/README.md [hunk 2]
```diff
@@ -14,3 +14,77 @@ appended to the end of the cluster-template.yaml, e.g cluster-template-{director
...
+{
+    "worker-flavors": ["default", "aks external-cloud-provider", "machinepool", "system-assigned-identity", "user-assigned-identity"]
+}
+```
+
+#### Requirements
+Please note your tilt-settings.json must contain at minimum the following fields:
+```json
+{
+    "kustomize_substitutions": {
```
Comment (CecileRobertMichon) [inline templates/flavors/README.md:36]: Isn't the minimum actually just the first 4 like documented in https://github.com/kubernetes-sigs/cluster-api-provider-azure/blob/master/docs/development.md#tilt-for-dev-in-capz?

```
{
  "kustomize_substitutions": {
        "AZURE_SUBSCRIPTION_ID_B64": "******",
        "AZURE_TENANT_ID_B64": "******",
        "AZURE_CLIENT_SECRET_B64": "******",
        "AZURE_CLIENT_ID_B64": "******",
  }
}
```

Comment (CecileRobertMichon) [reply templates/flavors/README.md:36]: Or is that a different requirement just when using tilt + flavors? If so, let's make that explicit, something like :

```
Please note your tilt-settings.json must contain at minimum the following fields when using tilt resources to deploy cluster flavors :
```

File: templates/flavors/README.md [hunk 3]
```diff
@@ -14,3 +14,77 @@ appended to the end of the cluster-template.yaml, e.g cluster-template-{director
 flavor can be used by specifying `--flavor {directory-name}`.
 
 To generate all CAPZ flavors, run `make generate-flavors`.
+
+
+## Running flavor clusters as a tilt resource
+
+#### From CLI
+run ```tilt up ${flavors}``` to spin up worker clusters in azure represented by a tilt local resource.  You can also modify flavors while tilt is up by running ```tilt args ${flavors}```
```
Comment (CecileRobertMichon) [inline templates/flavors/README.md:22]:
```suggestion
run ```tilt up ${flavors}``` to spin up worker clusters in Azure represented by a Tilt local resource.  You can also modify flavors while Tilt is up by running ```tilt args ${flavors}```
```

File: templates/flavors/README.md [hunk 4]
```diff
@@ -14,3 +14,77 @@ appended to the end of the cluster-template.yaml, e.g cluster-template-{director
...
+
+#### From Tilt Config
+Add your desired flavors to tilt_config.json:
+```json
+{
+    "worker-flavors": ["default", "aks external-cloud-provider", "machinepool", "system-assigned-identity", "user-assigned-identity"]
```
Comment (CecileRobertMichon) [inline templates/flavors/README.md:28]:
```suggestion
    "worker-flavors": ["default", "aks", "external-cloud-provider", "machinepool", "system-assigned-identity", "user-assigned-identity"]
```

File: templates/flavors/README.md [hunk 5]
```diff
@@ -14,3 +14,77 @@ appended to the end of the cluster-template.yaml, e.g cluster-template-{director
...
+        "AZURE_TENANT_ID": "******",
+        "AZURE_CLIENT_ID": "******",
+        "AZURE_CLIENT_SECRET": "******",
+        "AZURE_SSH_PUBLIC_KEY": "********"
+    }
+}
+```
+
+#### Defining Variable Overrides
+If you wish to override the default variables for flavor workers-- you can specify them as part of your tilt-settings.json as seen in the example below.  Please note-- precedence of variables is as follows:
```
Comment (CecileRobertMichon) [inline templates/flavors/README.md:51]:
```suggestion
If you wish to override the default variables for flavor workers, you can specify them as part of your tilt-settings.json as seen in the example below.  Please note, the precedence of variables is as follows:
```

File: templates/flavors/README.md [hunk 6]
```diff
@@ -14,3 +14,77 @@ appended to the end of the cluster-template.yaml, e.g cluster-template-{director
...
+
+1. explicitly defined vars for each flavor i.e. worker-templates.flavors[0].AZURE_VNET_NAME
```
Comment (CecileRobertMichon) [inline templates/flavors/README.md:53]: will the instructions in https://github.com/kubernetes-sigs/cluster-api-provider-azure/blob/master/docs/development.md#deploying-a-workload-cluster continue to work for using env vars to configure a workload cluster config if not using flavor as resources?

File: templates/flavors/README.md (hunk 2 above)
Comment (jroden) [reply templates/flavors/README.md:36]: Yes, the non B64 ones are only required when using flavors.  Good idea-- I will add that!

File: templates/flavors/README.md (hunk 6 above)
Comment (jroden) [reply templates/flavors/README.md:53]: Shouldn't be any interference between the two methods!  The flavors deployed via tilt will pull user-defined variables from tilt-settings.json (if present) whereas `make create-workload-cluster ` pulls them from environment variables

**Comment by jroden** [comment]
@CecileRobertMichon  thanks for the help! 
Once we're done with the review stage should I rebase all these commits into one?

**Comment by CecileRobertMichon** [comment]
/ok-to-test

@jroden please add a release note in the PR description and rebase + squash the commits (the bot won't squash merge so the commit history should look clean before merging).

File: Tiltfile [hunk 7]
```diff
@@ -208,3 +274,14 @@ if settings.get("deploy_cert_manager"):
 deploy_capi()
 
 capz()
+
+# run worker clusters specified from 'tilt up' or in 'tilt_config.json'
```
Comment (devigned) [inline Tiltfile:278]: Please wrap the lines below into a function. Perhaps, `flavors()`.

Also, please add a newline at the end of the file.

Comment (jroden) [reply Tiltfile]: Good catch!

**Comment by jroden** [comment]
Just rebased changes into one commit!

**Comment by jroden** [comment]
> /ok-to-test
> 
> @jroden please add a release note in the PR description and rebase + squash the commits (the bot won't squash merge so the commit history should look clean before merging).

Just added release note!

**Pull request merged**